### PR TITLE
Integration with Devise gem

### PR DESCRIPTION
**Pull request by lisset-orozco**
**Social-Red**
- [x] Authentication with Devise gem
- [x] Authentication with Facebook
- [x] Update data before registration
- [x] Styles with Material Desing Lite
- [x] Routes for authenticated and unauthenticated
- [x] Cycles and Interpolation in SASS
- [x] Implement navigation with Material Desing Lite

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+/tmp/*
+!/log/.keep
+!/tmp/.keep
+
+# Ignore Byebug command history file.
+.byebug_history
+
+#
+.rvmrc
+.env
```
Comment (joalbertg) [inline .gitignore]:
```suggestion
.env

```

File: Gemfile [hunk 2]
```diff
@@ -35,9 +35,22 @@ gem 'jbuilder', '~> 2.5'
 # Use Capistrano for deployment
 # gem 'capistrano-rails', group: :development
 
+# views 
+gem 'haml-rails', '~> 1.0'
+# Authentication
+gem 'devise', '~> 4.5'
+
+
```
Comment (joalbertg) [inline Gemfile]: delete line

File: Gemfile [hunk 3]
```diff
@@ -35,9 +35,22 @@ gem 'jbuilder', '~> 2.5'
...
 group :development, :test do
```
Comment (joalbertg) [inline Gemfile:46]: should in order alphabet the gems name

File: Gemfile [hunk 4]
```diff
@@ -35,9 +35,22 @@ gem 'jbuilder', '~> 2.5'
...
+  # configuration of local environment variables
+  gem 'dotenv-rails'
+  # console
+  gem 'awesome_print', '~> 1.8'
+  # debug
+  gem 'pry-nav', '~> 0.2.4'
+  gem 'pry-rails'
 end
 
 group :development do
```
Comment (joalbertg) [inline Gemfile:58]: the same

File: app/controllers/main_controller.rb [hunk 5]
```diff
@@ -0,0 +1,4 @@
+class MainController < ApplicationController
+  def home
```
Comment (joalbertg) [inline app/controllers/main_controller.rb]: in a line empty methods

File: app/views/main/home.haml [hunk 6]
```diff
@@ -0,0 +1 @@
+%h1 Bienvenido a la página
```
Comment (joalbertg) [inline app/views/main/home.haml]:
```suggestion
%h1 Bienvenido a la página

```

File: config/database.yml [hunk 7]
```diff
@@ -29,10 +29,10 @@ development:
   # To create additional roles in postgres see `$ createuser --help`.
   # When left blank, postgres will use the default role. This is
   # the same name as the operating system user that initialized the database.
-  username: Admin
+  username: <%= ENV['DATABASE_USER'] || 'postgres' %>
```
Comment (joalbertg) [inline config/database.yml:32]: update README file with information the environment variables

File: config/environments/development.rb [hunk 8]
```diff
@@ -51,4 +51,8 @@
   # Use an evented file watcher to asynchronously detect changes in source code,
   # routes, locales, etc. This feature depends on the listen gem.
   config.file_watcher = ActiveSupport::EventedFileUpdateChecker
+
+  # devise
+  config.action_mailer.default_url_options = { host: 'localhost', port: 3000 }
```
Comment (joalbertg) [inline config/environments/development.rb]: change by environment variables

File: config/locales/devise.en.yml [hunk 9]
```diff
@@ -0,0 +1,64 @@
+# Additional translations at https://github.com/plataformatec/devise/wiki/I18n
```
Comment (joalbertg) [inline config/locales/devise.en.yml:1]: add location .es file

**Review by joalbertg — CHANGES_REQUESTED** [review]
please verify the comments!!!

File: Gemfile (hunk 2 above)
Comment (lisset-orozco) [reply Gemfile]: line deleted

File: Gemfile (hunk 3 above)
Comment (lisset-orozco) [reply Gemfile:46]: ordered gems except the version of the rails for visibility

File: Gemfile (hunk 4 above)
Comment (lisset-orozco) [reply Gemfile:58]: ordered gems except the version of the rails for visibility

File: app/controllers/main_controller.rb (hunk 5 above)
Comment (lisset-orozco) [reply app/controllers/main_controller.rb]: method in a line

File: app/views/main/home.haml (hunk 6 above)
Comment (lisset-orozco) [reply app/views/main/home.haml]: line jump applied

File: config/locales/devise.en.yml (hunk 9 above)
Comment (lisset-orozco) [reply config/locales/devise.en.yml:1]: added file in Spanish

File: config/environments/development.rb (hunk 8 above)
Comment (lisset-orozco) [reply config/environments/development.rb]: variables already added

File: config/database.yml (hunk 7 above)
Comment (lisset-orozco) [reply config/database.yml:32]: Information added to the README

Comment (lisset-orozco) [reply config/database.yml:32]: Information added to the README

File: config/environments/development.rb (hunk 8 above)
Comment (lisset-orozco) [reply config/environments/development.rb]: variables already added

**Review by joalbertg — APPROVED** [review]

File: .gitignore [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+# Ignore all logfiles and tempfiles.
+/log/*
+/tmp/*
+!/log/.keep
+!/tmp/.keep
+
+# Ignore Byebug command history file.
+.byebug_history
+
+#
```
Comment (joalbertg) [inline .gitignore]: complete comment

File: app/controllers/users/omniauth_callbacks_controller.rb [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+        sign_in_and_redirect(@user, event: :authentication)
+      else
+        session['devise.auth'] = request.env['omniauth.auth']
+        render :edit
+      end
+    end
+
+    def custom_sign_up
+      @user = User.from_omniauth(session['devise.auth'])
+      if @user.update(user_params)
```
Comment (joalbertg) [inline app/controllers/users/omniauth_callbacks_controller.rb]: syntax in a line

File: app/controllers/users/omniauth_callbacks_controller.rb [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        sign_in_and_redirect(@user, event: :authentication)
+      else
+        render :edit
+      end
+    end
+
+    def failure
+      redirect_to new_user_session_path, notice: "Hubo un error con el login. Intenta de nuevo"
```
Comment (joalbertg) [inline app/controllers/users/omniauth_callbacks_controller.rb]: add parentheses to method

File: app/controllers/users/omniauth_callbacks_controller.rb [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    end
+
+    private
+
+    def user_params
+      # Strong Params
+      params.require(:user).permit(:name, :username, :email)
```
Comment (joalbertg) [inline app/controllers/users/omniauth_callbacks_controller.rb]: add line break between symbols

File: app/views/main/home.haml [hunk 14]
```diff
@@ -0,0 +1,2 @@
...
+= link_to "Cerrar sesión", destroy_user_session_path, method: :delete
```
Comment (joalbertg) [inline app/views/main/home.haml]: add parentheses to method

File: app/views/main/unregistered.haml [hunk 15]
```diff
@@ -0,0 +1,7 @@
+.full-height.row.middle-xs.white-text.center-xs.main-background.no-margin{style:"background-color:grey;"}
+  = grid xs:10, sm:8, md:6, lg:4 do
+    .big-padding.light-blue.text-left.white-text
+      %h1.no-margin.bold.light-green-text Red Social
+      %p.no-margin.bold Conecta y conoce a la comunidad
+      .top-space.no-margin
+        = render 'devise/registrations/new'
```
Comment (joalbertg) [inline app/views/main/unregistered.haml]: add parentheses to method

File: app/views/partials/_nav.haml [hunk 16]
```diff
@@ -0,0 +1,15 @@
+%header.mdl-layout__header.mdl-layout__header--scroll
+  .mdl-layout__header-row
+    / Title 
+    %span.mdl-layout-title Red Social
+    / Add spacer, to align navigation to the right 
+    .mdl-layout-spacer
+    / Navigation
+    %nav.mdl-navigation
+      -if user_signed_in?
+        = link_to "Cerrar sesión", destroy_user_session_path, method: :delete, class: "white-text"
```
Comment (joalbertg) [inline app/views/partials/_nav.haml]: add parentheses to method

File: app/views/users/omniauth_callbacks/edit.haml [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+    = text_field_tag "user[username]", @user.username
+    / <input type="text" name="user[username]" value="@user.username">
+  %div
+    %label{for:"user_email"} Correo electrónico
+    = email_field_tag "user[email]", @user.email, required: true
+  %div
+    %label{for:"user_name"} Nombre
+    = text_field_tag "user[name]", @user.name
+  %div
+    = submit_tag "Completar registro"
```
Comment (joalbertg) [inline app/views/users/omniauth_callbacks/edit.haml]: add line break

**Review by joalbertg — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 10 above)
Comment (lisset-orozco) [reply .gitignore]: Done

File: app/controllers/users/omniauth_callbacks_controller.rb (hunk 11 above)
Comment (lisset-orozco) [reply app/controllers/users/omniauth_callbacks_controller.rb]: Done

File: app/controllers/users/omniauth_callbacks_controller.rb (hunk 12 above)
Comment (lisset-orozco) [reply app/controllers/users/omniauth_callbacks_controller.rb]: Done!

File: app/controllers/users/omniauth_callbacks_controller.rb (hunk 13 above)
Comment (lisset-orozco) [reply app/controllers/users/omniauth_callbacks_controller.rb]: Done!

File: app/views/main/home.haml (hunk 14 above)
Comment (lisset-orozco) [reply app/views/main/home.haml]: Done!

File: app/views/main/unregistered.haml (hunk 15 above)
Comment (lisset-orozco) [reply app/views/main/unregistered.haml]: Done!

File: app/views/partials/_nav.haml (hunk 16 above)
Comment (lisset-orozco) [reply app/views/partials/_nav.haml]: Done!

File: app/views/users/omniauth_callbacks/edit.haml (hunk 17 above)
Comment (lisset-orozco) [reply app/views/users/omniauth_callbacks/edit.haml]: Done!

**Review by joalbertg — APPROVED** [review]
Done!!!

**Pull request merged**